[tcp] add rewrite of TCPlp's interface code to support OpenThread's TCP API

samkumar opened the pull request:
This is the ninth and final pull request in a series of pull requests (after #6744, #6770, #6790, #6857, #6885, #6925, #6926, and #7074) bringing TCPlp to OpenThread (see #6456). I'm making a series of small pull requests so that it's easier for the OpenThread community to review the code.

This pull request contributes support for the OpenThread TCP API (see #6411 and #6491) using the FreeBSD-derived functions from #7074. It is a rewrite of TCPlp's interface layer to support OpenThread's TCP API instead of the API used previously by TCPlp. Once this pull request is merged, the OpenThread CLI's TCP utility will work, and most of the OpenThread TCP API will work.

This pull request completes the sequence of pull requests bringing TCPlp to OpenThread. In previous pull requests, I pointed to #6650 for a preview of a working version of TCPlp in OpenThread. At this point, however, all functionality in #6650 is either merged into OpenThread's `main` branch or is included in this pull request. Therefore, #6650 is now obsolete and should be closed once this pull request is merged. To preview a working version of TCPlp in OpenThread, you should simply use the code in this pull request. I've done some testing and it seems to work fine. If you would like to try out TCP in OpenThread using the code in this pull request, the video I posted in #6650 (https://youtu.be/ppZ784YUKlI) might help you get started.

The remaining API features that are not yet supported, even with this pull request, include:
* TCP Fast Open
* `otTcpReceiveContiguify`

These features were not present in the original TCPlp code, so I feel it is most appropriate to contribute support for these features in other pull requests, separate from the pull requests contributing support for TCPlp.

samkumar commented:
One issue that I'd like to point out that will need to be resolved at some point (either in this pull request or in a future pull request) has to do with how FreeBSD supports TCP on big endian systems. Take a look at `struct tcphdr`, which corresponds to the TCP header: https://github.com/openthread/openthread/blob/0e8b6604cb256cef828a0ebb37bdd031cdf0677d/third_party/tcplp/bsdtcp/tcp.h.

The structure uses a C bit field to represent the "offset" field of the TCP header. The original FreeBSD code use compile-time `#if` checks to choose the ordering of the bit fields based on the endianness of the target platform. In the original TCPlp code, I commented out this logic and hardcoded the little endian choice, because the platforms I was using for experiments were little endian platforms. However, it seems that OpenThread aims to support both big endian and little endian devices, so we should fix this.

I can think of three ways to fix it. I'm fine with implementing any of these, but I'd like the OpenThread community's input to decide which one is most preferable:
1. In `core/net/tcp6.cpp`, when a packet is received, dynamically check if the system is big endian, and if it is, "fix up" the bit fields to be correct. For example, run `header.th_off = header.th_x2;`. This would happen after the checksum is verified. When packets are sent, "fix up" the bit fields to be correct. For example, you would read the corresponding fields out of the `Message` and write them in the correct order. This would happen before the checksum is computed.
2. Add logic similar to `core/common/endian.hpp` to check if the platform is big endian at compile time to `tcp.h`, and restore the endianness check in the FreeBSD code.
3. Do away with the bit fields in the definition of `struct tcphdr`. Instead, replace `th_off` and `th_x2` with a single field `th_off_x2`, and use bit operations to extract the offset when reading and writing the TCP header. These fields (`th_off` and `th_x2`) are only accessed a handful of times, so it would not require extensive changes to rewrite accesses to these fields with accesses to the new `th_off_x2` fields with the appropriate bit operations.

Personally, I prefer option number 3. The reason is that the C standard does not actually specify how bit fields are packed into bytes, which implies that what FreeBSD does with bit fields in the TCP header is not portable. For example, NetBSD acknowledges this directly in their code (see https://github.com/NetBSD/src/blob/7ea5251b473093b2a04e6003dea848e89cf6e2ee/sys/netinet/tcp.h). FreeBSD relies on the compiler's implementation details in order for this to work correctly, even with their `#if` checks, which may not be appropriate for OpenThread. Only option 3, which does away with the bit fields, solves this problem.

That said, I would like the OpenThread community's input on this issue, and I am open to other ideas. Once we decide on a fix, I am happy to either add it to this pull request, or open a new pull request with the fix.

samkumar commented:
I managed to make the CI tests all pass, but this required changing some of the Makefiles (`Android.mk` and `tests/unit/Makefile.am`) to properly link against the TCPlp module. Please take a look at those files as well, in case there is a simpler/better way to make the changes.

samkumar commented:
Regarding support for big endian platforms (see my comment above), I just added a new commit to this pull request implementing the third option. If others prefer one of the other options, I'm happy to remove this commit and implement the preferred fix.

jwhui commented:
> Personally, I prefer option number 3. The reason is that the C standard does not actually specify how bit fields are packed into bytes, which implies that what FreeBSD does with bit fields in the TCP header is not portable.

I prefer option 3 as well, for the same stated reason.

jwhui commented:
@samkumar , thanks for submitting this PR. Great to see we're at the end of this PR series!

Any chance you can add a CI test to check that TCP is functional?

samkumar commented:
I'm happy to add CI tests for TCP. It may take me a few days to do it because I'm currently not familiar with how CI tests in OpenThread work, and I'll need to build some familiarity with it in order to determine how best to write such a test.

Would you like the test to be part of this pull request, or should I submit it as a separate pull request?

jwhui commented:
> I'm happy to add CI tests for TCP. It may take me a few days to do it because I'm currently not familiar with how CI tests in OpenThread work, and I'll need to build some familiarity with it in order to determine how best to write such a test.

Sounds good. Thanks!

> Would you like the test to be part of this pull request, or should I submit it as a separate pull request?

I don't have a real preference as long as we get the CI tests in.

samkumar commented:
OK. In that case, I'll work on the CI tests. If I finish before this PR is reviewed, then I'll include it in this PR. Otherwise, I'll submit it as a separate PR.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -95,40 +125,69 @@ const SockAddr &Tcp::Endpoint::GetPeerAddress(void) const
 {
-    OT_UNUSED_VARIABLE(aSockName);
+    Error error;
+    VerifyOrExit(!static_cast<const Address *>(&aSockName.mAddress)->IsUnspecified(), error = kErrorInvalidArgs);
```
Newline after vairable definitions and first statement?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -95,40 +125,69 @@ const SockAddr &Tcp::Endpoint::GetPeerAddress(void) const
 
-    return kErrorNotImplemented;
+    memcpy(&mTcb.laddr, &aSockName.mAddress, sizeof(mTcb.laddr));
+    mTcb.lport = HostSwap16(aSockName.mPort);
```
I suggest adding 
```c++
using ot::Encoding::BigEndian::HostSwap16;
using ot::Encoding::BigEndian::HostSwap32;
```
at the top of this file to ensure we are using the big-endian here.

Without this I think we are indirectly relying on the `using` from `ip6_header.hpp`. I would say it may be better to be explicit about it at the top of the file.

samkumar replied:
OK, I added those `using` statements.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -95,40 +125,69 @@ const SockAddr &Tcp::Endpoint::GetPeerAddress(void) const
+
+    VerifyOrExit(mTcb.t_state == TCP6S_CLOSED, error = kErrorInvalidState);
+
+    struct sockaddr_in6 sin6p;
```
Move variable declaration to the top of scope?

From our style guide:

> - Scoped Variable declarations
>   - All scoped (i.e. stack) variable declarations should be placed together at the top of the enclosing scope in which they are used.
>   - There shall be an empty line after all such variable declarations.
>   - The names of all variable declarations should be left aligned.

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -95,40 +125,69 @@ const SockAddr &Tcp::Endpoint::GetPeerAddress(void) const
-    return kErrorNotImplemented;
+    bool moreToCome = (aFlags & OT_TCP_SEND_MORE_TO_COME) != 0;
+    int  bsdError   = tcp_usr_send(&mTcb, moreToCome ? 1 : 0, &aBuffer, 0);
+    SuccessOrExit(error = BsdErrorToOtError(bsdError));
```
newline?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -95,40 +125,69 @@ const SockAddr &Tcp::Endpoint::GetPeerAddress(void) const
+    SuccessOrExit(error = BsdErrorToOtError(bsdError));
+
+exit:
+    return error;
```
Simplify to this line?

```c++
return BsdErrorToOtError(tcp_user_send(&mTcp, (aFlags & OT_TCP_SEND_MORE_TO_COME) !=0, &aBuffer, 0);
```

samkumar replied:
I simplified the code to a single line as you suggested.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -163,25 +232,48 @@ Error Tcp::Endpoint::Deinitialize(void)
+    case TT_DELACK:
+        return kTimerDelack;
+    case TT_REXMT:
+        /* fallthrough intentional */
```
Use `OT_FALL_THROUGH` instead (to avoid some toolchain warnings/error)?

Or actually in this case, it may be fine to use
```
  case TT_REXMT:
  case TT_PERSIST:
       ...
```

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -163,25 +232,48 @@ Error Tcp::Endpoint::Deinitialize(void)
+        return kTimer2Msl;
+    default:
+        OT_ASSERT(false);
+        return 0; /* prevents a compiler warning */
```
Avoid using multiple `return` statements?

Per our style guide, we need a single `return` from function/method.

> 
>  - There should be one return statement per free function or method at the end of the free function or method.

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -163,25 +232,48 @@ Error Tcp::Endpoint::Deinitialize(void)
+    case kTimer2Msl:
+        return tcp_timer_active(&mTcb, TT_2MSL);
+    default:
+        return false;
```
Avoid multiple `return` statements? (single `return` at the end)

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -242,8 +334,38 @@ bool Tcp::Endpoint::FirePendingTimers(TimeMilli aNow, bool &aHasFutureTimer, Tim
+                 * times out).
+                 */
+                int dropped;
+                switch (timerIndex)
```
newline?

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+
+    VerifyOrExit(length == aMessage.GetLength() - aMessage.GetOffset(), error = kErrorParse);
+    VerifyOrExit(length >= sizeof(Tcp::Header), error = kErrorParse);
+    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset() + offsetof(struct tcphdr, th_off_x2), &headerSize, 1));
```
Use `sizeof(headersize)` instead of `1`?

abtink replied:
Actully, instead of 
```c++
Error Message::Read(uint16_t aOffset, void *aBuf, uint16_t aLength) const;
```
you can use the template version of `Message::Read<ObjectType>` to read a `uint8_t` value directly.
```c++
    template <typename ObjectType> Error Read(uint16_t aOffset, ObjectType &aObject) const
```
The code will be simpler (the compiler should infer the `ObjectType` to be `uint8_t` from the type of `headerSize`):
```c++
SuccessOrExit(error = aMessage.Read(aMessage.GetOffset() + offsetof(struct tcphdr, th_off_x2), headerSize));
```

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
-        OT_UNUSED_VARIABLE(active);
+        struct signals  sig;
+        otLinkedBuffer *priorHead = lbuf_head(&endpoint->mTcb.sendbuf);
+        memset(&sig, 0x00, sizeof(sig));
```
newline?

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+        struct signals  sig;
+        otLinkedBuffer *priorHead = lbuf_head(&endpoint->mTcb.sendbuf);
+        memset(&sig, 0x00, sizeof(sig));
+        int nextAction = tcp_input(ip6Header, tcpHeader, &aMessage, &endpoint->mTcb, nullptr, &sig);
```
Move with other vairable at the top of scope?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+        {
+            otLinkedBuffer *next = curr->mNext;
+            curr->mNext          = nullptr;
+            aEndpoint.mSendDoneCallback(&aEndpoint, curr);
```
We are calling user's callback here. Whenever we invoke a callback we need to consider that the user may call other functions/method to change stuff before we return from the callback. Is this possible here? Anything we need to check/validate after returning from callback.

In general, I would suggest to try to invoke the callback as the last statement (but I know there are situations when it may not be possible).

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+    case 0:
+        return kErrorNone;
+    default:
+        return kErrorFailed;
```
Single `return` statement?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+bool Tcp::CanBind(const SockAddr &aSockName)
+{
+    uint16_t port = HostSwap16(aSockName.mPort);
+    for (Endpoint *endpoint = mEndpoints.GetHead(); endpoint != nullptr; endpoint = endpoint->GetNext())
```
newline?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+            }
+            if (memcmp(&endpoint->mTcb.laddr, &aSockName.mAddress, sizeof(endpoint->mTcb.laddr)) == 0)
+            {
+                return false;
```
single `return` statement?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+            InstanceLocator::GetInstance().Get<Ip6>().SelectSourceAddress(peerInfo);
+        if (netifAddress == nullptr)
+        {
+            return false;
```
single return statement? (can use `ExitNow()`)

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+        MessageInfo peerInfo;
+        peerInfo.Clear();
+        peerInfo.SetPeerAddr(aPeer.GetAddress());
+        const Netif::UnicastAddress *netifAddress =
```
Variable declaration at the top of scope?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -395,4 +744,191 @@ void Tcp::ProcessTimers()
+        addr.mPort = HostSwap16(aPort);
+
+        otTcpEndpoint *               endpoint;
+        otTcpIncomingConnectionAction action = listener.mAcceptReadyCallback(&listener, &addr, &endpoint);
```
Variable at the top of scope?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -395,4 +744,191 @@ void Tcp::ProcessTimers()
+        case OT_TCP_INCOMING_CONNECTION_ACTION_DEFER:
+            return nullptr;
+        case OT_TCP_INCOMING_CONNECTION_ACTION_REFUSE:
+            return (struct tcpcb *)-1;
```
single `return`?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -395,4 +744,191 @@ void Tcp::ProcessTimers()
+void tcplp_sys_connection_lost(struct tcpcb *aTcb, uint8_t aErrNum)
+{
+    Tcp::Endpoint &endpoint = Tcp::Endpoint::FromTcb(*aTcb);
+    if (endpoint.mDisconnectedCallback != nullptr)
```
newline?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -395,4 +744,191 @@ void Tcp::ProcessTimers()
+                        bool              aBindAddress,
+                        bool              aBindPort)
+{
+    Instance &instance = *static_cast<Instance *>(aInstance);
```
Use `AsCoreType()`?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.hpp:
```diff
@@ -329,7 +332,22 @@ class Tcp : public InstanceLocator, private NonCopyable
+        static Endpoint &FromTcb(struct tcpcb &aTcb)
+        {
+            uintptr_t ptr = reinterpret_cast<uintptr_t>(&aTcb);
+            ptr -= offsetof(Endpoint, mTcb);
```
Is there a way to avoid using `offsetof()`?

I would avice avoid using it in general. It requires PoD types and can lead to some tricky bugs/issue (some toolchains have weird/unexpected behaviors).

abtink commented on src/core/net/tcp6.hpp:
```diff
@@ -438,6 +464,22 @@ class Tcp : public InstanceLocator, private NonCopyable
+        static Listener &FromTcbListen(struct tcpcb_listen &aTcbListen)
+        {
+            uintptr_t ptr = reinterpret_cast<uintptr_t>(&aTcbListen);
+            ptr -= offsetof(Listener, mTcbListen);
```
Same here about `offsetof`?

abtink commented on src/core/net/tcp6.hpp:
```diff
@@ -438,6 +464,22 @@ class Tcp : public InstanceLocator, private NonCopyable
+
+    private:
+        Address &      GetLocalIp6Address() { return *reinterpret_cast<Address *>(&mTcbListen.laddr); }
+        const Address &GetLocalIp6Address() const { return *reinterpret_cast<const Address *>(&mTcbListen.laddr); }
```
use `(void)` insetad of `()`?

samkumar replied:
OK. Fixed in the latest update.

abtink reviewed (commented):
Thank Sam. Overall looks good.
Some smaller suggestion below. Thanks.

samkumar commented:
Thanks @abtink for looking at the code so thoroughly in your review. I was able to directly address your comments about style and spacing by updating the code. There were three other comments that I want to discuss here:

## Including the relative path `../third_party/..` in `tcp.h`

It is necessary to include this file because of the `mTcb` field of `otTcpEndpoint`. The type of this field is `struct tcpcb`, a structure that is defined in the FreeBSD code. The key constraint that led to me doing this is that code-derived from third-party sources (e.g., code derived from FreeBSD that comes with its own license) needs to be in the `third_party` directory. This means that we can't solve this problem by moving the relevant structure definitions into `include/openthread`: they *must* remain in `third_party`.

If we *really* need to avoid the relative path in the `#include`, then we instead define the `mTcb` field as an opaque type --- a `struct` or `union` of the same size and alignment as `struct tcpcb`, and then use `reinterpret cast` in `tcp6.cpp`. The downside is that the size and alignment of this field would have to be kept in sync with `struct tcpcb`. We could potentially use a `static_assert` in `tcp6.cpp` to enforce this. What do you think of this solution? I am happy to implement this if you feel it is preferable to the current setup, which `#include`s a relative path.

## Using `offsetof` to convert `struct tcpcb *` to `Tcp::Endpoint *`
In the current code I believe it is used safely because all types for which I use `offsetof` are POD data types. If we `really` want to avoid using `offsetof`, then we can consider making `mTcb` the first field of `otTcpEndpoint` and then using a reinterpret cast without `offsetof` to convert from a pointer to one to a pointer to the other. This works because the C language, I believe, guarantees that no padding exists at the beginning of a structure (e.g., see https://stackoverflow.com/questions/142016/c-c-structure-offset).

Another possibility would be to add a `void *` field to `struct tcpcb` that points to the enclosing `Tcp::Endpoint`. I prefer the above solution to this one because I believe that this one has no clear advantages to the above approach but adds some memory overhead.

## Calling user callbacks
In the current code, I avoid calling user callbacks directly from the FreeBSD packet processing code, precisely to avoid the pathologies that you are thinking of (e.g., consider the case where we call user code when we are in the middle of processing a received TCP segment --- if the user code tries to send out data on that same TCP socket, the send routine will be called while the packet is in an inconsistent state). The mechanism that is currently implemented is that user callbacks are not called during TCPlp packet processing code. Instead, the callbacks to call and arguments to call them with are collected in `struct signals`. Then, once packet processing is complete, and `struct tcpcb` is in a consistent state, we finally look at `struct signals` and call the necessary callbacks. That's what `ProcessSignals` is doing. And note how, after `ProcessSignals` is called, we immediately call `ExitNow()`.

That said, looking at this carefully I realized that there is still one case that could be problematic --- if the user _deinitializes_ a `Tcp::Endpoint` in one of the user callbacks. Unfortunately, there is no way for `ProcessSignals` to detect that this happened, short of traversing the list of `Tcp::Endpoint`s to make sure that the `Tcp::Endpoint` for which signals are being processed has not been deinitialized.

Here are few options to fix this:
* Actually traverse the list of `Tcp::Endpoints` every time a user callback is called, in case the user deinitialized it.
* Specify in the documentation that the user must not call `otTcpEndpointDeinitialize` from a callback for that same TCP endpoint.

We can also consider restructuring the code to have a separate Tasklet from which all user callbacks are called. This would not directly solve any of the above issues, but it might make things easier to reason about. Some downsides are that this would require `struct signals` to be stored in `struct tcpcb`, adding memory overhead, and require the Tasklet callback to traverse all `Tcp::Endpoint`s, which adds CPU overhead.

samkumar commented:
To clarify my intent with the above post, I'm happy to implement any of the solutions I proposed (or any new suggestions beyond what I could think of), but I would like some guidance on how to proceed with those issues before attempting to address them.

samkumar commented:
@abtink I went ahead and implemented a solution for each of the three issues I had left unresolved at the time of my previous post. This is what I did.
1. I made `mTcb` and `mTcbListen` opaque types in `tcp.h`, allowing the third-party `#include` to be removed. I use `static_assert` in `tcp6.hpp` to check that the size and alignment of these opaque types match the size and alignment of the corresponding types in TCPlp.
2. I moved `mTcb` and `mTcbListen` to the top of `otTcpEndpoint` and `otTcpListener`, thereby eliminating the need to use `offsetof` in the functions you pointed out. I did add `static_assert`s to check that `mTcb` and `mTcbListen` have zero offset, but they aren't strictly necessary and I'm happy to remove them if you would like. I felt that using `offsetof` in the `static_assert`s was acceptable because any strange/compiler-dependent anomalies would result in a build failure, rather than silently causing the code to behave incorrectly.
3. Whenever user callbacks are called, I ensure that the corresponding `Endpoint` or `Listener` has not been deinitialized. I check this by searching for the corresponding `Endpoint` or `Listener` in the linked list of initialized `Endpoint`s or `Listener`s.

I hope that I've addressed all of your concerns now. Please take another look when you get a chance. Let me know if you'd prefer that I apply different solutions to address your concerns.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -242,8 +368,39 @@ bool Tcp::Endpoint::FirePendingTimers(TimeMilli aNow, bool &aHasFutureTimer, Tim
+                }
+                if (dropped != 0)
+                {
+                    return true;
```
Use `ExitNow()` instead of `return`?

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -395,4 +832,235 @@ void Tcp::ProcessTimers()
+    IgnoreError(instance.Get<ot::Ip6::Ip6>().SendDatagram(message, info, kProtoTcp));
+}
+
+uint32_t tcplp_sys_get_ticks()
```
```suggestion
uint32_t tcplp_sys_get_ticks(void)
```

samkumar replied:
OK. Fixed in the latest update.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -395,4 +832,235 @@ void Tcp::ProcessTimers()
+    return TimerMilli::GetNow().GetValue();
+}
+
+uint32_t tcplp_sys_get_millis()
```
```suggestion
uint32_t tcplp_sys_get_millis(void)
```

samkumar replied:
OK. Fixed in the latest update.

abtink reviewed (approved):
Thanks @samkumar. I was just reading the earlier comments (was off this week 😃 ). 
I think the 3 solutions you added already are perfect (what I would have suggested as well). Thanks.

It LGTM 👍 (just couple of smaller style suggestion below).

samkumar commented:
Thanks Abtin for taking another look. I just pushed an update including your suggested style fixes.

jwhui commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +497,271 @@ Error Tcp::Listener::Deinitialize(void)
+bool Tcp::CanBind(const SockAddr &aSockName)
+{
+    uint16_t port    = HostSwap16(aSockName.mPort);
+    bool     allowed = true;
```
nit: I think we could reduce a bit of code by:

- Initialize `allowed` to `false`.
- Remove `allowed = false` from `ExitNow()` instances below.
- Add `allowed = true;` before the `exit:` label.

Thoughts?

jwhui commented on src/core/net/tcp6.cpp:
```diff
@@ -395,4 +834,235 @@ void Tcp::ProcessTimers()
+    Tcp &          tcp      = listener.GetInstance().Get<Tcp>();
+    struct tcpcb * rv       = (struct tcpcb *)-1;
+
+    if (listener.mAcceptReadyCallback != nullptr)
```
nit: use `VerifyOrExit()` or `ExitNow()` to reduce indent below?

samkumar replied:
Done; implemented in the latest update.

jwhui commented on include/openthread/tcp.h:
```diff
@@ -610,14 +616,17 @@ typedef void (*otTcpAcceptDone)(otTcpListener *aListener, otTcpEndpoint *aEndpoi
-    void *                mContext;  ///< A pointer to application-specific context
+    union
+    {
+        uint8_t mSize[16 + 3 * sizeof(void *)];
```
Avoid magic numbers and comment on how this size was chosen?

samkumar replied:
OK. In the latest update, I use `#define`s for these constants, preceded by a comment stating how the size was chosen (to match `struct tcpcb_listen` in this case). I did the same for the opaque `mTcb` field in `otTcpEndpoint`.

jwhui commented on src/core/net/tcp6.cpp:
```diff
@@ -68,67 +77,133 @@ Error Tcp::Endpoint::Initialize(Instance &aInstance, otTcpEndpointInitializeArgs
+    {
+        uint8_t *recvbuf    = static_cast<uint8_t *>(aArgs.mReceiveBuffer);
+        size_t   recvbuflen = aArgs.mReceiveBufferSize - ((aArgs.mReceiveBufferSize + 8) / 9);
+        uint8_t *reassbmp   = recvbuf + recvbuflen;
+        lbuf_init(&tp->sendbuf);
```
```suggestion
        uint8_t *reassbmp   = recvbuf + recvbuflen;

        lbuf_init(&tp->sendbuf);
```

samkumar replied:
Fixed in the latest update.

jwhui commented on src/core/net/tcp6.cpp:
```diff
@@ -302,27 +497,271 @@ Error Tcp::Listener::Deinitialize(void)
+    }
+    if (aEndpoint.mSendDoneCallback != nullptr)
+    {
+        otLinkedBuffer *curr = aPriorHead;
+        for (int i = 0; i != aSignals.links_popped; i++)
```
```suggestion
        otLinkedBuffer *curr = aPriorHead;

        for (int i = 0; i != aSignals.links_popped; i++)
```

samkumar replied:
OK, fixed in the latest update.

jwhui commented on src/core/net/tcp6.hpp:
```diff
@@ -329,7 +332,32 @@ class Tcp : public InstanceLocator, private NonCopyable
+        /**
+         * Checks if this Endpoint is in the closed state.
+         */
+        bool IsClosed(void) { return GetTcb().t_state == TCP6S_CLOSED; }
```
```suggestion
        bool IsClosed(void) const { return GetTcb().t_state == TCP6S_CLOSED; }
```

samkumar replied:
Thanks. This is fixed in the latest update.

jwhui commented on src/core/net/tcp6.hpp:
```diff
@@ -546,13 +628,43 @@ class Tcp : public InstanceLocator, private NonCopyable
+    /**
+     * Automatically selects a local address and/or port for communication with the specified peer.
+     *
+     * @param[in] aPeer         The peer's address and port.
+     * @param[in,out] aToBind   The SockAddr into which to store the selected address and/or port.
+     * @param[in] aBindAddress  If true, the local address is selected; if not, the current address in @p aToBind is
+     * treated as a given.
+     * @param[in] aBindPort     If true, the local port is selected; if not, the current port in @p aToBind is treated
+     * as a given.
```
nit: fix alignment

samkumar replied:
Thanks. This is fixed in the latest update.

samkumar commented:
Thanks @jwhui for taking a look. I just pushed a commit addressing your comments.

samkumar commented:
I just rebased this onto the `main` branch to make sure everything still works OK.

simonlingoogle commented on src/core/net/tcp6.cpp:
```diff
@@ -68,67 +77,131 @@ Error Tcp::Endpoint::Initialize(Instance &aInstance, otTcpEndpointInitializeArgs
 Instance &Tcp::Endpoint::GetInstance(void)
 {
-    return AsCoreType(mInstance);
+    struct tcpcb *tp = &GetTcb();
+
+    return AsCoreType(tp->instance);
```
```suggestion
    struct tcpcb &tp = &GetTcb();

    return AsCoreType(tp.instance);
```

nit: prefer alias to pointer?

samkumar replied:
I made this change throughout the file (changing all `struct tcpcb *tp = &GetTcb();` into `struct tcpcb &tp = GetTcb();` and all uses of `tp` as needed).

simonlingoogle reviewed (approved):
LGTM :+1: 

with nits

samkumar commented:
Thanks @simonlingoogle for reviewing this pull request. I've updated it to address your comments.

I also rebased this onto `main` (again) and squashed it down to three commits. I didn't want to combine everything into a single commit, since the three commits do different things and it might be useful to retain that history.

@jwhui, at this point I think it is ready to merge. Let me know if anything else is necessary for that.

jwhui reviewed (approved):
Thanks! 👍🏼

jwhui commented:
@samkumar , great to reach this milestone!
